aten/src/ATen/functorch/BatchRulesBinaryOps.cpp

```diff
@@ -287,6 +287,24 @@ void fill__Tensor_batch_rule(
   std::get<0>(self_and_other).copy_(std::get<1>(self_and_other));
 }
 
+std::tuple<Tensor,optional<int64_t>> fill_Tensor_batch_rule(
+    Tensor& self,
+    optional<int64_t> self_bdim,
+    const Tensor& other,
+    optional<int64_t> other_bdim) {
+  if (!other_bdim.has_value()) {
+    // Optimization: fill is faster than the other path which does
+    // reshaping + copy_
+    return std::make_tuple(at::fill(self, other), self_bdim);
```

Comment:
maybe a dumb q: as you mentioned below, `at::fill()` requires `other` to be a scalar tensor. But in the vmap rule, `other` is not a scalar tensor - it's a tensor of shape `(batch_size,)`. Won't that cause `at::fill` to error?

Reply:
Let's say we're doing a single vmap. Not all tensors need to be vmapped over (only some may need to be). e.g.:
```
self = torch.randn(B)
other = torch.randn([])
vmap(torch.fill, (0, None))(self, other)
```
In this case only self gets wrapped in a BatchedTensor, while other is a regular Tensor. We could have always expanded `other` and wrapped it in a BatchedTensor, but that leads to performance problems: `other.expand(B) * 2` is slower than `other * 2` because PyTorch isn't very smart about strides.

In this example, fill_Tensor_batch_rule gets passed:
- self (tensor of shape [B])
- self_bdim (0)
- other (tensor of shape [])
- other_bdim (is nullopt -- that's what happens when other didn't get wrapped in a BatchedTensor)

And in the code you linked, we are calling `fill` on `self` and `other` only when other_bdim is nullopt.

Nested vmap is more complicated to reason through -- the batching rule is written with a single level of vmap in mind, but it automatically works on multiple levels of vmap. Let's say we have the following:
```
self = torch.randn(B0, B1)
other = torch.randn(B0)
vmap_0(vmap_1(torch.fill, (0, None)))(self, other)
```
We dispatch on vmap_1 first, and fill_Tensor_batch_rule gets passed:
- self (BatchedTensor wrapping Tensor of shape [B0, B1])
- self_bdim (0)
- other (BatchedTensor wrapping Tensor of shape [B0])
- other_bdim (is nullopt -- `other` isn't getting vmapped over in vmap_1)
`at::fill(self, other)` is still legit here -- other.sizes() returns []

Reply:
ah. Thanks for explaining the nested case!